heat/core/arithmetics.py

```diff
@@ -197,6 +203,179 @@ def bitwise_xor(t1, t2):
+
+        recv = recv.reshape(cprod.shape[:axis] + torch.Size([1]) + cprod.shape[axis + 1 :])
+        cprod *= recv
+
+    if out is not None:
+        return out
+
+    return factories.array(
+        cprod, dtype=a.dtype if dtype is None else dtype, is_split=a.split, device=a.device
+    )
+
+
+# Alias support
+cumproduct = cumprod
+
+
+def cumsum(a, axis, dtype=None, out=None):
+    """
+    Return the cumulative sum of the elements along a given axis.
+
+    Parameters
+    ----------
+    a : DNDarray
+        Input array.
+    axis : int
+        Axis along which the cumulative sum is computed.
+    dtype : dtype, optional
+        Type of the returned array and of the accumulator in which the
+        elements are summed.  If `dtype` is not specified, it defaults
+        to the dtype of `a`, unless `a` has an integer dtype with a
+        precision less than that of the default platform integer.  In
+        that case, the default platform integer is used.
+    out : DNDarray, optional
+        Alternative output array in which to place the result. It must
+        have the same shape and buffer length as the expected output
+        but the type will be cast if necessary. See `doc.ufuncs`
+        (Section "Output arguments") for more details.
+
+    Returns
+    -------
+    cumsum_along_axis : DNDarray.
+        A new array holding the result is returned unless `out` is
+        specified, in which case a reference to `out` is returned. The
+        result has the same size as `a`, and the same shape as `a` if
+        `axis` is not None or `a` is a 1-d array.
+
+    """
+    if axis is None:
+        raise NotImplementedError("axis = None is not supported")
+    if not isinstance(axis, int):
+        raise RuntimeError("axis parameter got an invalid argument {}".format(axis))
+
+    if dtype is not None:
+        dtype = types.canonical_heat_type(dtype)
+
+    if out is not None:
+        if out.shape != a.shape:
+            raise RuntimeError(
+                "out and a have different shapes {} != {}".format(out.shape, a.shape)
+            )
+        if out.split != a.split:
+            raise RuntimeError(
+                "out and a have different splits {} != {}".format(out.split, a.split)
+            )
+        if out.device != a.device:
+            raise RuntimeError(
+                "out and a have different devices {} != {}".format(out.device, a.device)
+            )
+        dtype = out.dtype
+
+    csum = torch.cumsum(
+        a._DNDarray__array,
+        axis,
+        out=None if out is None else out._DNDarray__array,
+        dtype=None if dtype is None else dtype.torch_type(),
+    )
+
+    if a.split is not None and axis == a.split:
+        send = torch.zeros(csum.shape[:axis] + csum.shape[axis + 1 :], dtype=csum.dtype)
+
+        if csum.numel() > 0:
+            indices = -1
+            Ni, Nk = csum.shape[:axis], csum.shape[axis + 1 :]
+            for ii in np.ndindex(Ni):
```

Comment:
Can we somehow manage to do this without the Python loops via views? Performance on them should be slow.

Reply:
I replaced it with torch.index_select()